journal.py

```diff
@@ -0,0 +1,186 @@
+
+HERE = os.path.dirname(os.path.abspath(__file__))
+DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))
+DATABASE_URL = os.environ.get(
+    'DATABASE_URL',
+    'postgresql://jesse:Jjk5646!@localhost:5432/learning-journal'
+)
+
+
+Base = declarative_base()
+engine = sa.create_engine(DATABASE_URL)
+# Session = sessionmaker(bind=engine)
+
+
+class Entry(Base):
+    """Make a new entry
+    """
+    __tablename__ = 'entries'
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
+    title = sa.Column(sa.Unicode(127), nullable=False)
+    date = sa.Column(
+        sa.DateTime, nullable=False, default=datetime.datetime.utcnow
+    )
+    content = sa.Column(sa.UnicodeText, nullable=False)
+
+    @classmethod
+    def write(cls, title=None, content=None, session=None):
+        if session is None:
+            session = DBSession
+        instance = cls(title=title, content=content)
+        session.add(instance)
+        return instance
+
+    @classmethod
+    def all(cls, session=None):
+        if session is None:
+            session = DBSession
+        return session.query(cls).order_by(cls.date.desc()).all()
+
+
+def init_db():
+    """Make a new entries table
+    """
+    Base.metadata.create_all(bind=engine)
+
+
+@view_config(route_name='home', renderer='templates/index.jinja2')
+def list_view(request):
+    entries = Entry.all()
+    return {'entries': entries}
+
+
+@view_config(route_name='add', request_method='POST')
+def add_entry(request):
+    title = request.params.get('title')
+    content = request.params.get('content')
+    Entry.write(title=title, content=content)
+    return HTTPFound(request.route_url('home'))
+
+
+@view_config(context=DBAPIError)
+def db_exception(context, request):
+    from pyramid.response import Response
+    response = Response(context.message)
+    response.status_int = 500
+    return response
+
+
+@view_config(route_name='logout')
+def logout(request):
+    headers = forget(request)
+    return HTTPFound(request.route_url('home'), headers=headers)
+
+
+@view_config(route_name='login', renderer="templates/login.jinja2")
+def login(request):
+    """authenticate a user by username/password"""
+    username = request.params.get('username', '')
+    error = ''
+    if request.method == 'POST':
+        error = "Login Failed"
+        authenticated = False
+        try:
+            authenticated = do_login(request)
+        except ValueError as e:
+            error = str(e)
+
+        if authenticated:
+            headers = remember(request, username)
+            return HTTPFound(request.route_url('home'), headers=headers)
+
+    return {'error': error, 'username': username}
+
+
+@view_config(route_name='create', renderer="templates/create.jinja2")
+def create(request):
+    """go to create page"""
+    if not request.authenticated_userid:               # hackers get redirected
+        return HTTPFound(request.route_url('login'))   # to login
+    return {}
```

Comment:
good use of checking `request.authenticated_userid` to see if the user has "permission" to be here!